charger/vehicle_controlled_charger.go

```diff
@@ -0,0 +1,188 @@
+	geofenceEnabled  bool
+	chargerLatitude  float64
+	chargerLongitude float64
+	radius           float64
+}
+
+func init() {
+	registry.Add("vehicle-controlled", NewVehicleControlledChargerFromConfig)
+}
+
+// NewVehicleControlledChargerFromConfig creates a new vehicle-controlled charger
+func NewVehicleControlledChargerFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		GeofenceEnabled bool    `mapstructure:"geofence_enabled"`
+		Latitude        float64 `mapstructure:"latitude"`
+		Longitude       float64 `mapstructure:"longitude"`
+		Radius          float64 `mapstructure:"radius"`
+	}{
+		GeofenceEnabled: false,
+		Radius:          100, // Default 100 meter radius
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	c := &VehicleControlledCharger{
+		log:              util.NewLogger("vehicle-controlled"),
+		enabled:          false,
+		wasDisconnected:  true,
+		geofenceEnabled:  cc.GeofenceEnabled,
+		chargerLatitude:  cc.Latitude,
+		chargerLongitude: cc.Longitude,
+		radius:           cc.Radius,
+	}
+
+	return c, nil
+}
+
+// isVehicleAtCharger checks if the vehicle is within the geofence (if enabled)
+func (c *VehicleControlledCharger) isVehicleAtCharger(vehicle api.Vehicle) (bool, error) {
+	if !c.geofenceEnabled {
+		return true, nil // Assume at charger if geofencing is disabled
+	}
+
+	positioner, ok := vehicle.(api.VehiclePosition)
+	if !ok {
+		return false, errors.New("vehicle must support position tracking if geofence is enabled")
+	}
+
+	lat, lng, err := positioner.Position()
+	if err != nil {
+		return true, err
+	}
+
+	distance := simpleDistance(c.chargerLatitude, c.chargerLongitude, lat, lng)
+	return distance <= c.radius, nil
+}
+
+// Status implements the api.Charger interface
+func (c *VehicleControlledCharger) Status() (api.ChargeStatus, error) {
+	if c.lp == nil {
+		return api.StatusA, nil
+	}
+
+	vehicle := c.lp.GetVehicle()
+	if vehicle == nil {
+		return api.StatusA, nil // No vehicle = disconnected
+	}
+
+	// Check if vehicle is at the charger (trying to use geofencing)
+	vehicleIsAtCharger, err := c.isVehicleAtCharger(vehicle)
+	if err != nil {
+		c.log.ERROR.Printf("%v. Assuming vehicle is at charger", err)
+		vehicleIsAtCharger = true
+	}
+
+	chargeState, ok := vehicle.(api.ChargeState)
+	if !ok {
+		return api.StatusA, errors.New("vehicle not capable of reporting charging status")
+	}
+	vehicleAPIStatus, err := chargeState.Status()
+
+	if err != nil {
+		return api.StatusA, err
+	}
+
+	if vehicleAPIStatus == api.StatusA || !vehicleIsAtCharger {
+		c.wasDisconnected = true
+		c.enabled = false
+		return api.StatusA, nil
+	} else {
+		if c.wasDisconnected {
+			c.enabled = vehicleAPIStatus == api.StatusC
+			c.wasDisconnected = false
+		}
+
+		if c.enabled {
+			return api.StatusC, nil
+		} else {
+			return api.StatusB, nil
+		}
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (c *VehicleControlledCharger) Enabled() (bool, error) {
```

Comment:
Enabled should be if it is *allowed* to charge, not if it's actually charging.

Reply:
yes

Reply:
Okay. In order to address this, I have introduced the c.isCharging property, which is distinct from c.enabled.